source/Octopus.Tentacle/Kubernetes/PodLogReader.cs

```diff
@@ -9,41 +9,51 @@ namespace Octopus.Tentacle.Kubernetes
 {
     static class PodLogReader
     {
-        public static async Task<(IReadOnlyCollection<ProcessOutput> Lines, long NextSequenceNumber)> ReadPodLogs(long lastLogSequence, StreamReader reader)
+        public static async Task<(IReadOnlyCollection<ProcessOutput> Lines, long NextSequenceNumber, int? exitCode)> ReadPodLogs(long lastLogSequence, StreamReader reader)
         {
+            int? exitCode = null;
             var results = new List<ProcessOutput>();
             var nextSequenceNumber = lastLogSequence;
             bool haveSeenPodLogEntryMatchingLogSequence = false;
-            
             while (true)
             {
                 var line = await reader.ReadLineAsync();
 
                 //No more to read
                 if (line.IsNullOrEmpty())
                 {
-                    return (results, nextSequenceNumber);
+                    return (results, nextSequenceNumber, exitCode);
                 }
 
-                //TODO: print parse errors
                 var parseResult = PodLogLineParser.ParseLine(line!);
-                var podLogLine = parseResult.LogLine;
-                
-                //Pod log line numbers are 1-based, log sequence is 0-based
-                if (podLogLine != null && podLogLine.LineNumber > lastLogSequence)
+
+                if (parseResult is ValidPodLogLineParseResult validParseResult)
                 {
-                    if (podLogLine.LineNumber == lastLogSequence + 1)
-                        haveSeenPodLogEntryMatchingLogSequence = true;
+                    var podLogLine = validParseResult.LogLine;
 
-                    if (!haveSeenPodLogEntryMatchingLogSequence)
-                        throw new MissingPodLogException();
+                    //Pod log line numbers are 1-based, log sequence is 0-based
+                    if (podLogLine.LineNumber > lastLogSequence)
+                    {
+                        if (podLogLine.LineNumber == lastLogSequence + 1)
+                            haveSeenPodLogEntryMatchingLogSequence = true;
 
-                    results.Add(new ProcessOutput(podLogLine.Source, podLogLine.Message, podLogLine.Occurred));
-                    nextSequenceNumber = podLogLine.LineNumber;
-                }
+                        if (!haveSeenPodLogEntryMatchingLogSequence)
+                            throw new MissingPodLogException();
+
+                        if (validParseResult is EndOfStreamPodLogLineParseResult endOfStreamParseResult)
+                            exitCode = endOfStreamParseResult.ExitCode;
```

Comment:
if we are parsing the EOS message, should be just abort exit the loop here rather than waiting for the next read line? Should we read logs _after_ the EOS message?

Reply:
It's still possible for the Pod to error after this line if something goes wrong - I think we still want to show whatever comes after this message?